iroha/src/tx.rs

```diff
@@ -73,39 +73,53 @@ impl AcceptedTransaction {
         world_state_view: &WorldStateView,
         permissions_validator: &PermissionsValidatorBox,
         is_genesis: bool,
-    ) -> Result<ValidTransaction, String> {
+    ) -> Result<ValidTransaction, RejectedTransaction> {
         let mut world_state_view_temp = world_state_view.clone();
         let account_id = self.payload.account_id.clone();
         if !is_genesis && account_id == <Account as Identifiable>::Id::genesis_account() {
             return Err(
-                "Genesis account can sign only transactions in the genesis block.".to_string(),
+                self.reject("Genesis account can sign only transactions in the genesis block.")
             );
         }
         world_state_view
             .read_account(&account_id)
-            .ok_or(format!("Account with id {} not found", account_id))?
+            .ok_or_else(|| {
+                self.clone()
+                    .reject(&format!("Account with id {} not found", account_id))
+            })?
             .verify_signature(
-                self.signatures.first().ok_or("No signatures found.")?,
+                self.signatures
+                    .first()
+                    .ok_or_else(|| self.clone().reject("No signatures found."))?,
                 self.hash().as_ref(),
-            )?;
+            )
+            .map_err(|error| self.clone().reject(&error))?;
         for instruction in &self.payload.instructions {
             let account_id = self.payload.account_id.clone();
             world_state_view_temp = instruction
                 .clone()
-                .execute(account_id.clone(), &world_state_view_temp)?;
+                .execute(account_id.clone(), &world_state_view_temp)
+                .map_err(|error| self.clone().reject(&error))?;
             if !is_genesis {
-                permissions_validator.check_instruction(
-                    account_id.clone(),
-                    instruction.clone(),
-                    &world_state_view,
-                )?;
+                permissions_validator
+                    .check_instruction(account_id.clone(), instruction.clone(), &world_state_view)
+                    .map_err(|error| self.clone().reject(&error))?;
             }
         }
         Ok(ValidTransaction {
             payload: self.payload,
             signatures: self.signatures,
         })
     }
+
+    /// Rejects transaction with the `rejection_reason`.
+    pub fn reject(self, rejection_reason: &str) -> RejectedTransaction {
+        RejectedTransaction {
+            payload: self.payload,
+            signatures: self.signatures,
+            rejection_reason: rejection_reason.to_string(),
+        }
+    }
```

Comment:
Better to replace with constructor for `RejectedTransaction::from_...`.

Reply:
Why do you think?
`transaction.reject()` seems more intuitive and compact to me than
`RejectedTransaction::from_valid_transaction(transaction, rejection_reason)`.

Also all our pipeline of txs and blocks is organized in the way that we don't have constructors but instead *move the pipeline stage* from the previous entity. I think this is a nice feature.

Reply:
I see, ok - agree with your point.